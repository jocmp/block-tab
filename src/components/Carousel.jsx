import React from 'react';
import PropTypes from 'prop-types';
import _ from 'underscore';
import classNames from 'classnames';

class Carousel extends React.Component {
  constructor(props) {
    super(props);
    this.state = { canScrollLeft: false, canScrollRight: true };

    this.scrollToStart = this.scrollToStart.bind(this);
    this.scrollToEnd = this.scrollToEnd.bind(this);
    this.updateScrollOptions = this.updateScrollOptions.bind(this);
  }

  componentDidMount() {
    window.addEventListener('resize', this.updateScrollOptions);
    this.carousel.addEventListener('scroll', this.updateScrollOptions);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.updateScrollOptions);
    this.carousel.removeEventListener('scroll', this.updateScrollOptions);
  }

  updateScrollOptions() {
    this.setState({
      canScrollLeft: this.carousel.scrollLeft > 0,
      canScrollRight: this.carousel.scrollLeft != (this.carousel.scrollWidth - this.carousel.clientWidth)
    });
  }

  scrollToStart() {
    this.carousel.scrollBy({ left: -this.carousel.scrollWidth, behavior: 'smooth' });
    this.updateScrollOptions();
  }

  scrollToEnd() {
    this.carousel.scrollBy({ left: this.carousel.scrollWidth, behavior: 'smooth' });
    this.updateScrollOptions();
  }

  render() {
    return (
      <div className="carousel__wrapper">
        <div className="carousel" ref={ (carousel) => { this.carousel = carousel; } }>
          { this.props.tabs.map((tab, index) =>
            <span
              key={ index }
              onClick={ this.props.onTabClick(tab.id) }
              className={ classNames('carousel__tab', { 'carousel__tab--selected': tab.id === this.props.selectedTabId }) }>
              { tab.title }
            </span>
          ) }
        </div>
        <button
          className={ classNames('carousel__button', 'btn--left', { 'btn--more': this.state.canScrollLeft }) }
          onClick={ this.scrollToStart }>
          <i className="indicator fa fa-chevron-left" />
        </button>
        <button
          className={ classNames('carousel__button', 'btn--right', { 'btn--more': this.state.canScrollRight }) }
          onClick={ this.scrollToEnd }>
          <i className="indicator fa fa-chevron-right" />
        </button>
      </div>
    );
  }
}

export default Carousel;

Carousel.propTypes = {
  selectedTabId: PropTypes.string,
  onTabClick: PropTypes.func.isRequired,
  tabs: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string,
    title: PropTypes.string,
    text: PropTypes.string,
    media: PropTypes.string
  })).isRequired
};
