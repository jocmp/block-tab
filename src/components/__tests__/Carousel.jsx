import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Carousel from '../Carousel';
import { tabs } from './fixtures';

describe('Carousel', () => {

  it('renders correctly', () => {
    const wrapper = mount(<Carousel tabs={ tabs() } onTabClick={ jest.fn() } />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('responds to click events', () => {
    const onTabClick = jest.fn();
    const wrapper = mount(<Carousel tabs={ tabs() } onTabClick={ onTabClick } />);
    wrapper.find('.carousel__tab').first().simulate('click');
    expect(onTabClick).toHaveBeenCalled();
  });

  it('handles an empty tab', () => {
    const wrapper = mount(<Carousel tabs={ [...tabs(), {}] } onTabClick={ jest.fn() } />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});