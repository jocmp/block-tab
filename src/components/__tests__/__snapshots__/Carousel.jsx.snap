// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel handles an empty tab 1`] = `
<Carousel
  onTabClick={
    [MockFunction] {
      "calls": Array [
        Array [
          "0001",
        ],
        Array [
          "0010",
        ],
        Array [
          "0101",
        ],
        Array [
          undefined,
        ],
      ],
    }
  }
  tabs={
    Array [
      Object {
        "id": "0001",
        "text": "<p>Description here</p>",
        "title": "Title 1",
      },
      Object {
        "id": "0010",
        "text": "<p>Description here</p>",
        "title": "Title 2",
      },
      Object {
        "id": "0101",
        "media": "fake://img.con",
        "text": "<p>Description here</p>",
        "title": "Title 3",
      },
      Object {},
    ]
  }
>
  <div
    className="carousel__wrapper"
  >
    <div
      className="carousel"
    >
      <span
        className="carousel__tab"
        key="0"
      >
        Title 1
      </span>
      <span
        className="carousel__tab"
        key="1"
      >
        Title 2
      </span>
      <span
        className="carousel__tab"
        key="2"
      >
        Title 3
      </span>
      <span
        className="carousel__tab carousel__tab--selected"
        key="3"
      />
    </div>
    <button
      className="carousel__button btn--left"
      onClick={[Function]}
    >
      <i
        className="indicator fa fa-chevron-left"
      />
    </button>
    <button
      className="carousel__button btn--right btn--more"
      onClick={[Function]}
    >
      <i
        className="indicator fa fa-chevron-right"
      />
    </button>
  </div>
</Carousel>
`;

exports[`Carousel renders correctly 1`] = `
<Carousel
  onTabClick={
    [MockFunction] {
      "calls": Array [
        Array [
          "0001",
        ],
        Array [
          "0010",
        ],
        Array [
          "0101",
        ],
      ],
    }
  }
  tabs={
    Array [
      Object {
        "id": "0001",
        "text": "<p>Description here</p>",
        "title": "Title 1",
      },
      Object {
        "id": "0010",
        "text": "<p>Description here</p>",
        "title": "Title 2",
      },
      Object {
        "id": "0101",
        "media": "fake://img.con",
        "text": "<p>Description here</p>",
        "title": "Title 3",
      },
    ]
  }
>
  <div
    className="carousel__wrapper"
  >
    <div
      className="carousel"
    >
      <span
        className="carousel__tab"
        key="0"
      >
        Title 1
      </span>
      <span
        className="carousel__tab"
        key="1"
      >
        Title 2
      </span>
      <span
        className="carousel__tab"
        key="2"
      >
        Title 3
      </span>
    </div>
    <button
      className="carousel__button btn--left"
      onClick={[Function]}
    >
      <i
        className="indicator fa fa-chevron-left"
      />
    </button>
    <button
      className="carousel__button btn--right btn--more"
      onClick={[Function]}
    >
      <i
        className="indicator fa fa-chevron-right"
      />
    </button>
  </div>
</Carousel>
`;
